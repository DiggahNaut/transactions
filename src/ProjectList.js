import React from 'react';

const ProjectList = (props) => {
    var filteredProjectList = [];
    var transactionsData = props.transactionsData;

    // Если я правильно понял суть задачи построения списка проектов, тогда отобразятся те проекты, по которым "успешно" совершились транзакции.
    // Могу быть не прав, прошу прощения за неуточнение этого вопроса. Много времени не займёт на правку данного момента.
    const projectList = transactionsData.map((row, index) => {
        if (row.transaction.status !== 'canceled') {
            if (filteredProjectList.includes(row.transaction.project.name)) {
                return;
            }

            filteredProjectList.push(row.transaction.project.name);

            return <li className="list-group-item" key={index}> {row.transaction.project.name} </li>;
        }
    });

    return (
        <div className="text-center">
            <h3>Список проектов, по которым совершались платежи.</h3>
            <div>
                Список строится на основе успешных транзакций у проекта. Отмененные не учитываются.
                Порядок проектов ни на чём не основан.
            </div>

            <ul className="list-group">
                {projectList}
            </ul>
        </div>
    )
}

export default ProjectList;