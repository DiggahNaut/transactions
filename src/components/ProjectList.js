import React, {Component} from 'react';

class ProjectList extends Component {
    constructor(props) {
        super(props);

        this.state = {
            projectList: ''
        };
    }

    handleChange = (e) => {
        var status = e.target.value;

        this.setProjectList(status);
    };

    setProjectList(status) {
        var filteredProjectList = [];
        var projectList = '';
        
        projectList = this.props.transactionsData.map((row, index) => {
            if (row.transaction.status !== status && !filteredProjectList.includes(row.transaction.project.name)) {
                filteredProjectList.push(row.transaction.project.name);
    
                return <li className="list-group-item" key={index}> {row.transaction.project.name} </li>;
            }
        });
    
        this.setState({projectList: projectList});
    }

    render () {
        return (
            <div className="text-center">
                <h3>Список проектов, по которым совершались транзакции.</h3>
                <div>
                    Выберите интересующий статус транзакции и список проектов изменится на основе статуса.
                </div>

                <select name="project-statuses" className="form-control" onChange={this.handleChange} defaultValue="Статус транзакции">
                    <option disabled>
                        Статус транзакции
                    </option>

                    <option value="created">
                        Новая
                    </option>

                    <option value="done">
                        Успешная
                    </option>

                    <option value="canceled">
                        Отмененная
                    </option>
                </select>

                <ul className="list-group list">
                    {this.state.projectList}
                </ul>
            </div>
        );
    };
}

export default ProjectList;