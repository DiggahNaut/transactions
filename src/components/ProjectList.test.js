import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ProjectList from './ProjectList';
import transactionsData from '../transactions.json';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ProjectList transactionsData={transactionsData}/>, div);
});

it("check ProjectList snapshot", () => {
    const ProjectListComponent = renderer.create(<ProjectList transactionsData={transactionsData}/>).toJSON();
    expect(ProjectListComponent).toMatchSnapshot();
});

// Я так понял что в компоненте ProjectList можно протестировать изменения селектора.
// Но похоже что для этого нужны имитация браузера и имитация действий пользователя.
// Могу быть не прав и это можно сделать с помощью jest/enzyme. Но до меня не дошло как протестировать изменения в селекторе.
// Пробовал это сделать с помощью jest/enzyme. Порылся в интернете, ничего полезного не нашёл.
describe('checking statuses for ProjectList', () => {
    const wrapper = mount(<ProjectList transactionsData={transactionsData}/>);

    it('check amount of statuses', () => {
        expect(wrapper.find('option').length).toEqual(4);
    });
    
    // it('check default status', () => {
    //     expect(wrapper.find('option')).toBe('Статус транзакции')
    // });

    // it('status = created', () => {
    //     wrapper.find('select[name="project-statuses"]').simulate('change', {target: {name: 'project-statuses', value: 'created'}});
    //     expect(wrapper.find('select[name="project-statuses"]').props('value')).toBe('created');
    // });

    // it('status = done', () => {
    //     wrapper.find('select').simulate('change', {target: {value: 'done'}})
    //     expect(wrapper.find('select').props('option')).toBe('done')
    // });

    // it('status = canceled', () => {
    //     wrapper.find('select').simulate('change', {target: {value: 'canceled'}})
    //     expect(wrapper.find('option').value).toBe('canceled')
    // });
});