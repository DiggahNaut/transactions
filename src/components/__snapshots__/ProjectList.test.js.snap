// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check ProjectList snapshot 1`] = `
<div
  className="text-center"
>
  <h3>
    Список проектов, по которым совершались транзакции.
  </h3>
  <div>
    Выберите интересующий статус транзакции и список проектов изменится на основе статуса.
  </div>
  <select
    className="form-control"
    defaultValue="Статус транзакции"
    name="project-statuses"
    onChange={[Function]}
  >
    <option
      disabled={true}
    >
      Статус транзакции
    </option>
    <option
      value="created"
    >
      Новая
    </option>
    <option
      value="done"
    >
      Успешная
    </option>
    <option
      value="canceled"
    >
      Отмененная
    </option>
  </select>
  <ul
    className="list-group list"
  >
    
  </ul>
</div>
`;
